tools/updater/main.c

```diff
@@ -934,174 +862,220 @@ static k4a_result_t command_inspect_firmware(updater_command_info_t *command_inf
-    if (!K4A_SUCCEEDED(result))
+    for (uint32_t device_index = 0; device_index < command_info->device_count; device_index++)
     {
-        return result;
-    }
+        k4a_result_t result = ensure_firmware_open(command_info, device_index);
+        if (!K4A_SUCCEEDED(result))
+        {
+            return result;
+        }
 
-    // Load and parse the firmware file information...
-    result = command_inspect_firmware(command_info);
-    if (!K4A_SUCCEEDED(result))
-    {
-        return result;
-    }
+        // Make a copy of the command so that only 1 device is updated at a time.
+        updater_command_info_t sub_command_info = { 0 };
+        sub_command_info.device_count = 1;
+        sub_command_info.device_serial_number[0] = command_info->device_serial_number[device_index];
+        sub_command_info.firmware_handle = command_info->firmware_handle;
+        sub_command_info.firmware_serial_number = command_info->firmware_serial_number;
+        sub_command_info.firmware_path = command_info->firmware_path;
 
-    bool audio_current_version_same = compare_version(command_info->current_version.audio,
-                                                      command_info->firmware_package_info.audio);
-    bool depth_config_current_version_same =
-        compare_version_list(command_info->current_version.depth_sensor,
-                             command_info->firmware_package_info.depth_config_number_versions,
-                             command_info->firmware_package_info.depth_config_versions);
-    bool depth_current_version_same = compare_version(command_info->current_version.depth,
-                                                      command_info->firmware_package_info.depth);
-    bool rgb_current_version_same = compare_version(command_info->current_version.rgb,
-                                                    command_info->firmware_package_info.rgb);
-
-    printf(
-        "Please wait, updating device firmware. Don't unplug the device. This operation can take a few minutes...\n");
-
-    // Write the loaded firmware to the device...
-    result = firmware_download(command_info->firmware_handle,
-                               command_info->firmware_buffer,
-                               command_info->firmware_size);
-    if (!K4A_SUCCEEDED(result))
-    {
-        printf("ERROR: Downloading the firmware failed! %d\n", result);
-        return result;
-    }
+        // Query the current device information...
+        result = command_query_device(&sub_command_info);
+        if (!K4A_SUCCEEDED(result))
+        {
+            return result;
+        }
 
-    // Wait until the update operation is complete and query the device to get status...
-    bool update_failed = K4A_FAILED(wait_update_operation_complete(command_info->firmware_handle, &finalStatus));
+        // Load and parse the firmware file information...
+        result = command_inspect_firmware(&sub_command_info);
```

Comment:
This loads and inspects the firmware. This should get called outside the for loop to prevent leaking memory and in general to prevent duplicating loading from disk.